rfcs/0144-versioned-flake-references.md

```diff
@@ -0,0 +1,118 @@
+---
+feature: versioned-flake-references
+start-date: 2023-03-19
+author: figsoda
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Flake references can have a special placeholder that will specify the version
+requirement of the flake, which `nix flake update` can use to update the pin to
+the latest version compatible with the version specification. A new Nix
+command, `nix flake upgrade` will upgrade the version requirement in
+`flake.nix` to the latest possible version without regard to compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+This will allow Nix libraries to be versioned without requiring their users
+to manually update them. Some package managers (e.g. cargo) for more
+conventional programming languages have this functionality built-in, allowing
+library authors to introduce breaking changes in a communicatable way that will
+not break their downstream dependents's code.
+
+# Detailed design
+[design]: #detailed-design
+
+## Syntax
+Version placeholders will be in `{}`, with a list of version requirements
+separated by `,` (commas), e.g. `{^1.0,<3}` will expand to a version that
+specifies the version requirements `^1.0` and `<3`
+
+Version requirements consists of a comparator and version. The version does not
+have to be a valid version defiined by the versioning scheme, `1` and `1.0` are
+also valid versions for version requirements. The comparator has to be one of
+the following options:
+- `^` compatible (as defined by the versioning scheme)
+- `<` less than
+- `<=` less than or equal to
+- `>` greater than
+- `>=` greater than or equal to
+- `=` equal to
+
+A version placeholder with no version requirements will match all valid
+versions defined by the versioning scheme.
+
+## Versioning scheme
+The versions will follow [semantic versioning] (semver). The tags have to
+follow semver, `1.0` and `1.0.0.0` will not match the placeholder `{=1.0}` as
+it is not a valid version defined by semver. Build identifiers will not be
+taken into consideration when calculating version requirements, and the latest
+tag will be selected if multiple tags have the same precedence defined by
+semver.
+
+## Upgrading
+A new Nix command, `nix flake upgrade`, will edit `flake.nix` and upgrade all
+the version placeholders in the inputs. A new flag, `--upgrade-input` will
```

Comment:
Upgrading `flake.nix` in-place is not really an option since we don't have the ability to do that. But I also don't see why it's necessary. `nix flake update` can simply update to the latest versions that are consistent with the version placeholders. At most, we want a flag to `nix flake update` to allow/disallow major upgrades that wouldn't be semantically compatible.

Reply:
If a flag were to temporarily ignore semantic compatibility, wouldn't that introduce some weird behaviour? Wouldn't the `update` behaviour on subsequent uses (without the flag) be to *downgrade* to a compatible version, thus defeating the purpose of initially using the flag? The idea here is to permanently alter the semver to accept the latest version as a convenience (so the user doesn't need to track down the latest tags of *every* input). I can't see any way of doing that (without potentially confusing behaviour) but inplace changes to `flake.nix`.

Alternatively, we could have a command/flag which would report the latest tags of the inputs (similar to how `update` reports lock file changes). This way there's no need to edit `flake.nix` inplace and the user gets the necessary information at once instead of playing tag whack-a-mole. It's less powerful but still works to make version tracking much less annoying which, I think, is the ulterior motive of this RFC.